File: typescript/examples/mcp-enforcement.ts
/**
 * Example: Governed MCP tool invocation via KSwitchInterceptor.
 *
 * This is the SUPPORTED production pattern for gating MCP tool calls through
 * KSwitch governance policies. It uses the governed invocation path:
 *
 *   KSwitchInterceptor.checkAndInvoke() → local PDP → enforce → tool → output filter → audit
 *
 * The old pattern (client.enforcement.enforceMCPCall()) is deprecated and does
 * NOT provide local-PDP evaluation, bypass prevention, output filtering, or
 * obligation blocking. Use KSwitchInterceptor.checkAndInvoke() instead.
 *
 * Usage:
 *   npx tsx examples/mcp-enforcement.ts
 */

import {
  KSwitchClient,
  KSwitchInterceptor,
  KSwitchEnforcementError,
  KSwitchObligationError,
  OutputDeniedError,
  LocalPDPEvaluator,
} from "../src/index.js";

// ── Simulated MCP tool implementations ────────────────────────────────────────

async function readRecords(table: string, limit = 10): Promise<object> {
  return { records: Array.from({ length: limit }, (_, i) => ({ id: i, table })) };
}

async function deleteRecords(table: string, ids: number[]): Promise<object> {
  return { deleted: ids.length, table };
}

// ── Governed invocation (supported path) ──────────────────────────────────────

async function main() {
  const client = new KSwitchClient({
    baseUrl: process.env.KSWITCH_BASE_URL ?? "http://localhost:5001",
    clientId: process.env.KSWITCH_CLIENT_ID,
    clientSecret: process.env.KSWITCH_CLIENT_SECRET,
    keycloakUrl: process.env.KEYCLOAK_ENDPOINT,
  });

  // Create the interceptor with optional local PDP for in-process enforcement.
  // The interceptor is the primary governed invocation surface for TypeScript.
  const evaluator = new LocalPDPEvaluator();
  const interceptor = new KSwitchInterceptor(client, {
    localPDP: evaluator,
  });

  const agentId = "agent:fraud-detector@bank.internal";
  const mcpServerId = "mcp:database@bank.internal";

  console.log("Checking tool call authorizations via governed interceptor\n");

  // ── Governed invocations ───────────────────────────────────────────────────
  // checkAndInvoke() is the ONLY supported production path.
  // Local PDP evaluate → enforce → output filter → audit is automatic.

  const calls: Array<{ toolName: string; toolFn: (...args: any[]) => Promise<any>; args: object }> = [
    { toolName: "read_records", toolFn: readRecords, args: { table: "customers", limit: 10 } },
    { toolName: "delete_records", toolFn: deleteRecords, args: { table: "customers", ids: [1, 2, 3] } },
  ];

  for (const { toolName, toolFn, args } of calls) {
    console.log(`Invoking ${toolName}...`);
    try {
      const result = await interceptor.checkAndInvoke({
        agentId,
        mcpServerId,
        toolName,
        toolFn: () => (toolFn as any)(...Object.values(args)),
      });
      console.log(`  ALLOWED — result:`, result);
    } catch (err) {
      if (err instanceof KSwitchEnforcementError) {
        console.warn(`  DENIED — reason: ${err.reason}`);
      } else if (err instanceof KSwitchObligationError) {
        console.warn(`  BLOCKED (obligation) — reason: ${err.reason}`);
      } else if (err instanceof OutputDeniedError) {
        console.warn(`  OUTPUT DENIED — policy prevents export of this result`);
      } else {
        throw err;
      }
    }
    console.log();
  }
}

main().catch(console.error);

File: python/examples/mcp_enforcement.py
#!/usr/bin/env python3
"""Example: Governed MCP tool invocation via KSwitchRuntime.

This is the SUPPORTED production pattern for gating MCP tool calls through
KSwitch governance policies. It uses the governed invocation path:

    KSwitchRuntime.invoke() → enforce → tool execution → output filtering → audit

The old pattern (client.enforcement.enforce_mcp_call()) is deprecated and does
NOT provide bypass prevention, output filtering, or obligation blocking.
Use KSwitchRuntime.invoke() or KSwitchInterceptor.check_and_invoke() instead.

Usage:
    python mcp_enforcement.py
"""

from kswitch import (
    KSwitchClient,
    KSwitchRuntime,
    KSwitchEnforcementError,
    KSwitchObligationError,
    OutputDeniedError,
)


# ── Simulated MCP tool implementations ───────────────────────────────────────

def read_records(table: str, limit: int = 10) -> dict:
    """Simulated MCP tool: read database records."""
    return {"records": [{"id": i, "table": table} for i in range(limit)]}


def delete_records(table: str, ids: list) -> dict:
    """Simulated MCP tool: delete database records."""
    return {"deleted": len(ids), "table": table}


# ── Governed invocation (supported path) ─────────────────────────────────────

def main():
    client = KSwitchClient(
        base_url="https://localhost:5001",
        client_id="mcp-database-server",
        client_secret="mcp-secret",
        keycloak_url="http://keycloak:8080",
        verify_ssl=False,
    )

    # Create the governed runtime for this agent + MCP server.
    # The runtime is the primary public invocation surface.
    runtime = KSwitchRuntime(
        agent_id="agent:fraud-detector@bank.internal",
        mcp_server_id="mcp:database@bank.internal",
        client=client,  # enables server fallback + revocation sync + audit
    )

    # Register tools. runtime.register() returns a GovernedTool — the
    # governed callable that always routes through enforcement.
    runtime.register("read_records", read_records)
    runtime.register("delete_records", delete_records)

    print("Checking tool call authorizations via governed runtime\n")

    # ── Governed invocation ───────────────────────────────────────────────────
    # This is the ONLY supported production path.
    # enforce → output filter → audit is automatic — no manual wiring needed.

    calls = [
        ("read_records", {"table": "customers", "limit": 10}),
        ("delete_records", {"table": "customers", "ids": [1, 2, 3]}),
    ]

    for tool_name, kwargs in calls:
        print(f"Invoking {tool_name}...")
        try:
            result = runtime.invoke(tool_name, **kwargs)
            print(f"  ALLOWED — result: {result}")
        except KSwitchEnforcementError as e:
            print(f"  DENIED — reason: {e.args[0]}")
        except KSwitchObligationError as e:
            print(f"  BLOCKED (obligation) — reason: {e.reason}")
        except OutputDeniedError:
            print(f"  OUTPUT DENIED — policy prevents export of this result")
        print()

    # Print observability diagnostics
    print("Revocation sync:", runtime.revocation_sync_diagnostics()["sync_worker"]["running"])
    print("Audit forwarding:", runtime.audit_diagnostics()["forwarding_enabled"])

    runtime.stop_revocation_sync()
    runtime.stop_audit_sender()
    client.close()


if __name__ == "__main__":
    main()
